lib/proxy/log_buffer/decorator.rb

```diff
@@ -1,32 +1,53 @@
 require 'proxy/log_buffer/decorator'
 require 'proxy/log_buffer/buffer'
+require 'thread'
 
 module Proxy::LogBuffer
   class Decorator
+    def self.instance
+      @@instance ||= new(::Proxy::LoggerFactory.logger)
+    end
+
     def initialize(logger, buffer = Proxy::LogBuffer::Buffer.instance)
       @logger = logger
       @buffer = buffer
+      @mutex = Mutex.new
+      @log_rolled = false
+    end
+
+    # due to synchronization can't re-open the log from the signal trap
+    def log_rolled
+      @log_rolled = true
+    end
+
+    def handle_log_rolling
+      need_rolling = !@logger.instance_of?(::Syslog::Logger) rescue true
+      @logger = ::Proxy::LoggerFactory.logger if need_rolling
+      @log_rolled = false
     end
 
     def add(severity, message = nil, progname = nil, backtrace = nil)
-      severity ||= UNKNOWN
-      if message.nil?
-        if block_given?
-          message = yield
-        else
-          message = progname
+      @mutex.synchronize do
```

Comment:
Is this to prevent the old logger being written to during the rolling?

If so, we could perhaps remove it and allow the old logger to be written to for that brief period - httpd works this way and logrotate is then usually configured with `delaycompress`.  The idea's that the old log gets written to as the pre-rotation workers finish off their requests and the log only gets compressed on the next rotation.  The old file descriptor's still valid so loggers still work, it's just that the file gets moved to a new name.